src/compiler/types/generate-event-listener-types.ts

```diff
@@ -0,0 +1,80 @@
+import { dashToPascalCase } from '@utils';
+
+import type * as d from '../../declarations';
+import { updateTypeIdentifierNames } from './stencil-types';
+
+/**
+ * Generates event listener properties for the component's html element type. This is used to allow implementers
+ * to use strict typings when adding and removing event listeners.
+ *
+ * @param cmp The component compiler metadata
+ * @param typeImportData locally/imported/globally used type names, which may be used to prevent naming collisions
+ * @returns additional types information to add event listener method overloads for component's html element type
+ */
+export const generateEventListenerTypes = (
+  cmp: d.ComponentCompilerMeta,
+  typeImportData: d.TypesImportData,
+): { htmlElementEventMap: string[]; htmlElementEventListenerProperties: string[] } => {
+  const tagName = cmp.tagName.toLowerCase();
+  const tagNameAsPascal = dashToPascalCase(tagName);
+  const htmlElementName = `HTML${tagNameAsPascal}Element`;
+  const cmpEventInterface = `${tagNameAsPascal}CustomEvent`;
+  const htmlElementEventMapName = `${htmlElementName}EventMap`;
+  const cmpEvents = cmp.events.filter((cmpEvent) => cmpEvent.complexType.original);
+  if (!cmpEvents.length) {
+    return { htmlElementEventMap: [], htmlElementEventListenerProperties: [] };
+  }
+  return {
+    htmlElementEventMap: getHtmlElementEventMap(cmpEvents, typeImportData, cmp.sourceFilePath, htmlElementEventMapName),
+    htmlElementEventListenerProperties: [
+      `                addEventListener<K extends keyof ${htmlElementEventMapName}>(type: K, listener: (this: ${htmlElementName}, ev: ${cmpEventInterface}<${htmlElementEventMapName}[K]>) => any, options?: boolean | AddEventListenerOptions): void;`,
+      `                addEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | AddEventListenerOptions): void;`,
```

Comment:
I ran into an issue testing this against the Ionic Framework. They have a handful of cases where they use `addEventListener` and `removeEventListener`  with the following type signatures:
```ts
        addEventListener<K extends keyof DocumentEventMap>(type: K, listener: (this: Document, ev: DocumentEventMap[K]) => any, options?: boolean | AddEventListenerOptions): void;
```
```ts
    removeEventListener<K extends keyof HTMLElementEventMap>(type: K, listener: (this: HTMLElement, ev: HTMLElementEventMap[K]) => any, options?: boolean | EventListenerOptions): void;
```

Can you do me a favor and add those? The first of the two is in the suggestion below:
```suggestion
      `                addEventListener<K extends keyof DocumentEventMap>(type: K, listener: (this: Document, ev: DocumentEventMap[K]) => any, options?: boolean | AddEventListenerOptions): void;`,
      `                addEventListener(type: string, listener: EventListenerOrEventListenerObject, options?: boolean | AddEventListenerOptions): void;`,
```